lib/Sema/CSFix.cpp

```diff
@@ -1508,3 +1508,33 @@ bool SpecifyKeyPathRootType::diagnose(const Solution &solution,
   
   return failure.diagnose(asNote);
 }
+
+bool UnwrapOptionalBaseKeyPathApplication::diagnose(const Solution &solution,
+                                                    bool asNote) const {
+  MissingOptionalUnwrapKeyPathFailure failure(solution, getFromType(),
+                                              getToType(), getLocator());
+  return failure.diagnose(asNote);
+}
+
+UnwrapOptionalBaseKeyPathApplication *
+UnwrapOptionalBaseKeyPathApplication::attempt(ConstraintSystem &cs, Type baseTy,
+                                              Type rootTy,
+                                              ConstraintLocator *locator) {
+  if(baseTy->hasTypeVariable() || rootTy->hasTypeVariable())
+    return nullptr;
+
+  if (!isExpr<SubscriptExpr>(locator->getAnchor()))
+    return nullptr;
+  
+  // Only diagnose this if base is an optional type and we only have a
+  // single level of optionality so we can safely suggest unwrapping.
+  auto nonOptionalTy = baseTy->getOptionalObjectType();
+  if (!nonOptionalTy || nonOptionalTy->getOptionalObjectType())
+    return nullptr;
+  
+  if (!TypeChecker::isSubtypeOf(nonOptionalTy, rootTy, cs.DC))
```

Comment:
Can't use `isSubtypeOf` here because that creates a separate constraint system, please use existing `cs` and `matchTypes` and check whether it's a subtype since it's known that both types are "concrete".